### PR TITLE
Erase extension types eagerly during translation of type symbols to Cci entities.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Relates to test plan https://github.com/dotnet/roslyn/issues/66722

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@jcouv FYI

**Comment by jcouv** [comment]
I'm going to rebase (from [this commit](https://github.com/dotnet/roslyn/pull/74576/commits/44beb47093325faa13744a8576c5ef61c5a616c7), for the record) to resolve conflicts now that the baseline was updated/merged.

File: src/Compilers/CSharp/Portable/Symbols/ExtensionTypeEraser.cs [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+            if (!changed)
+            {
+                return type;
+            }
+
+            return type.WithTypeAndModifiers(transformedType, type.CustomModifiers);
+        }
+
+        private static NamedTypeSymbol TransformNamedType(NamedTypeSymbol type, out bool changed)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/ExtensionTypeEraser.cs:87]: I'd rename this to `TransformExcludingSelf`.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1). Will push a small PR to address two nits

File: src/Compilers/Core/Portable/PEWriter/ReferenceIndexerBase.cs [hunk 2]
```diff
@@ -15,12 +15,8 @@ namespace Microsoft.Cci
 {
     internal abstract class ReferenceIndexerBase : MetadataVisitor
     {
-        // Note: for the _alreadySeen set, we use the KeepExtensions flag to differentiate between
-        // a visit that erases extension types from one that doesn't.
-        // For example, for `C<Extension>` we'll need to collect references from the erased type `C<ExtendedType>`
-        // but also references needed to decode the extension erasure attribute `[ExtensionErasure("C'1[Extension]")]`.
-        private readonly HashSet<IReferenceOrISignature> _alreadySeen = new HashSet<IReferenceOrISignature>();
-        private readonly HashSet<IReferenceOrISignature> _alreadyHasToken = new HashSet<IReferenceOrISignature>();
+        private readonly HashSet<IReferenceOrISignature> _alreadySeen = new();
```
Comment (jjonescz) [inline src/Compilers/Core/Portable/PEWriter/ReferenceIndexerBase.cs:18]: I think we disallow `new()` in compiler code.

File: src/Dependencies/PooledObjects/ArrayBuilder.cs [hunk 3]
```diff
@@ -569,6 +569,14 @@ public void AddRange<U>(ArrayBuilder<U> items, Func<U, T> selector)
             }
         }
 
+        public void AddRange<U>(ImmutableArray<U> items, Func<U, T> selector)
+        {
```
Comment (jjonescz) [inline src/Dependencies/PooledObjects/ArrayBuilder.cs:573]: Should we add something like the following?

```cs
EnsureCapacity(Count + items.Length);
```

**Review by jjonescz — APPROVED** [review]

File: src/Dependencies/PooledObjects/ArrayBuilder.cs (hunk 3 above)
Comment (jcouv) [reply src/Dependencies/PooledObjects/ArrayBuilder.cs:573]: Yup, that seems good. I'll leave the other overloads as-is though.

**Pull request merged**